streams/src/main/java/org/apache/kafka/streams/processor/TaskMetadata.java

```diff
@@ -40,6 +40,18 @@
 
     private final Optional<Long> timeCurrentIdlingStarted;
 
+    /**
+     * @deprecated since 3.0, please use {@link #TaskMetadata(TaskId, Set, Map, Map, Optional) instead}
+     */
+    @Deprecated
+    public TaskMetadata(final String taskId,
+                        final Set<TopicPartition> topicPartitions,
+                        final Map<TopicPartition, Long> committedOffsets,
+                        final Map<TopicPartition, Long> endOffsets,
+                        final Optional<Long> timeCurrentIdlingStarted) {
+        this(TaskId.parse(taskId), topicPartitions, committedOffsets, endOffsets, timeCurrentIdlingStarted);
+    }
+
     public TaskMetadata(final TaskId taskId,
```

Comment:
Should we make this protected and add a internal "impl" class? (And also mark as non-public so we can eventually move to an interface?)

Reply:
Well, we would still need a public constructor no matter what even with an internal impl class. I'll add a comment to clarify that it's not intended for public use and maybe file a followup ticket to migrate this to an interface in case someone wants to get into that

Reply:
https://issues.apache.org/jira/browse/KAFKA-12849

Reply:
> Well, we would still need a public constructor

Why can't we make it `protected`.

Reply:
I meant it seems like we should need it as long as TaskMetadata continues to be a public class, ie if we take away the constructor then we've essentially all but made it into an interface already -- which was not what was agreed on in KIP-740, and I don't want to drag this out further if there is any pushback...

Imo it's sufficient to just leave the constructor as `public` but not present it as a public API, and leave the rest for the KIP for [KAFKA-12849](https://issues.apache.org/jira/browse/KAFKA-12849). It's a simple KIP so I'm optimistic someone new to kafka may want to pick it up 🙂 

All that said I don't feel too strongly (in fact I agree with you) so I can be convinced if you do

Reply:
In fact we already got a bite on  KAFKA-12849 -- if it's all the same to you then I'd prefer to just let this be handled by that KIP. Assuming we get it in by 3.0 then there won't ever be a new public constructor published for TaskMetadata

Reply:
Works for me. Thanks for the details.